app/controllers/application_controller.rb

```diff
@@ -5,6 +5,140 @@ class ApplicationController < Sinatra::Base
+      redirect '/tweets', locals: {message: "You're already logged in!"}
+    end
+  end
+
+  post "/signup" do
+    user = User.new(username: params[:username], email: params[:email], password: params[:password])
+    if user.save
+      session[:user_id] = user.id
+      redirect '/tweets', locals: {message: "Sign up successful, please log in now."}
+    else
+      redirect '/signup', locals: {message: "Sign up failed, try again."}
+    end
+  end
+
+  get "/login" do
+    if !logged_in?
+      erb :'/users/login'
+    else
+      redirect '/tweets', locals: {message: "You're already logged in!"}
+    end
+  end
+
+  post "/login" do
+    user = User.find_by(username: params[:username])
+    if user && user.authenticate(params[:password])
+      session[:user_id] = user.id
+      redirect "/tweets", locals: {message: "Successfully logged in."}
+    else
+      redirect "/login", locals: {message: "Log in failed, try again."}
+    end
+  end
+
+  get "/logout" do
+    if logged_in?
+      session.clear
+      redirect "/login", locals: {message: "Successfully logged out."}
+    else
+      redirect "/login", locals: {message: "Can't log out, you're not logged in."}
+    end
+  end
+
+  get "/users/:slug" do
+    @user = User.find_by_slug(params[:slug])
+    erb :'/users/show'
+  end
+
+  get "/tweets" do
+    if logged_in?
+      @tweets = Tweet.all
+      erb :'/tweets/tweets'
+    else
+      redirect "/login", locals: {message: "Please log in to see that."}
+    end
+  end
+
+  get "/tweets/new" do
+    if logged_in?
+      erb :'/tweets/new'
+    else
+      redirect "/login", locals: {message: "Please log in to see that."}
+    end
+  end
+
+  post "/tweets" do
+    if logged_in? && params[:content] != ""
+      Tweet.create(content: params[:content], user_id: session[:user_id])
+    elsif params[:content] == ""
+      redirect "/tweets/new", locals: {message: "Tweets can't be empty."}
+    else
+      redirect "/login", locals: {message: "Please log in to see that."}
+    end
+  end
+
+  get "/tweets/:id" do
+    if logged_in?
+      @tweet = Tweet.find(params[:id])
+      erb :'/tweets/show'
+    else
+      redirect "/login", locals: {message: "Please log in to see that."}
+    end
+  end
+
+  get "/tweets/:id/edit" do
+    if logged_in?
+      @tweet = Tweet.find(params[:id])
+      erb :'/tweets/edit'
+    else
+      redirect "/login", locals: {message: "Please log in to see that."}
+    end
+  end
+
+  post "/tweets/:id" do
+    @tweet = Tweet.find(params[:id])
+    if logged_in? && params[:content] != ""
+      @tweet.update(content: params[:content])
+      redirect "/tweets/#{@tweet.id}"
+    elsif params[:content] == ""
+      redirect "/tweets/#{@tweet.id}/edit", locals: {message: "Tweets can't be empty."}
+    else
+      redirect "/login", locals: {message: "Please log in to see that."}
+    end
+  end
+
+  delete "/tweets/:id/delete" do
+    @tweet = Tweet.find(params[:id])
+    if logged_in? && @tweet.user == current_user
```

Comment:
Prefer to find tweets that are already only associated to the `current_user` than to manually check the `user_id` foreign key of the tweet against the `current_user.id`

```ruby
tweet = current_user.tweets.find_by(:id => params[:id])
if tweet && tweet.destroy
  redirect "/tweets"
else
  redirect "/tweets/#{tweet.id}"
end
```

We're using the association collection method provided by [`has_many`](http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_many) to query the user's tweets. 

Note the usage of [`find_by`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find-by) vs [`find`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find). Had we used `find` here, ActiveRecord would raise an error in the event that we could not find a tweet associated with this user. `find_by` will allow this action to fail gracefully, simply returning nil if the user is trying to delete a tweet that they do not have access to.